lua/caching_handlers.lua

```diff
@@ -166,10 +166,18 @@ function caching_handlers._parse_request(incoming_zipkin_headers)
     local destination = spectre_common.get_smartstack_destination(request_headers)
     local cacheability_info = spectre_common.determine_if_cacheable(normalized_uri, destination, request_headers)
 
-    -- Remove the gzip header because it's easier to work with text responses
-    if ngx.re.match(ngx.req.get_headers()['accept-encoding'], 'gzip') then
-        ngx.req.clear_header("accept-encoding")
-        request_headers['accept-encoding'] = nil
+    -- Modify request headers if it's a bulk endpoint since we need the response to
+    -- be pure uncompressed JSON
+    if cacheability_info.is_cacheable and cacheability_info.cache_entry.bulk_support then
+        -- Remove the gzip header because it's easier to work with text responses
+        if ngx.re.match(ngx.req.get_headers()['accept-encoding'], 'gzip') then
+            ngx.req.clear_header("accept-encoding")
+            request_headers['accept-encoding'] = nil
+        end
+        -- Let's also set the 'application/json' Accept header since we can only
+        -- handle pure json responses.
+        ngx.req.set_header('accept', 'application/json')
```

Comment:
Can we not just pop msgpack? I feel like we might lose something this way. I don't know what though, it just feels less good. 

For instance, some of our servers actually send `Content-Type: application/json; charset=utf-8`, would some proxy decide this isn't a match and drop it?

Where the bravado clients support adding the msgpack header, would it be `Accept: */*, application/msgpack`? Or 2 distinct lines of accept headers? Can openresty easily handle it that way instead?

Reply:
Well, we only support json so I'd either enforce json or drop the header entirely. Sending through `Accept: */*` is wrong as we only want json back.

Reply:
Dropping `application/msgpack` is harder than expected. `ngx.req.get_headers()['accept']` returns different results depending on how you send the headers.

With 2 or more separate accept headers (i.e. `-H 'Accept: application/json' -H 'accept: application/msgpack'`) returns a list of them: `["application/json","application/msgpack"]`.

If I only set one header, it only returns a single string:
- `-H 'Accept: application/json, application/msgpack'` --> `"application/json, application/msgpack"`
- `-H 'Accept: application/msgpack'` --> `"application/msgpack"`

For all 3 cases, I'd then need to split each string on `,` and figure out if in any part there's something similar to `application/msgpack`, drop only that, reconstruct the original string / array and override the request headers.

Possible, but seems overcomplicated and error prone. Given that msgpack encoding is something people have to enable explicitly, I'd just drop this entire PR and add a note to the docs saying that for bulk endpoints we don't support msgpack encoding. (Normal endpoints will work fine). cc @thebostik

Reply:
Ok I'm with you, I don't have a concern and I think it's fair to just document this is how it works. Just want to check whether we want to "reset" the request header value to `application/json` still, because it's always `*/*` right now. It might make more sense to reset it to the latter value since for our use case, since we know what request header everything sends, and that's effectively the same as popping off `application/msgpack` for us.

Reply:
Oh ok, what about the accept-encoding change?

Let's assume that users will still mess it up, and enable spectre proxying and send the accept encoding header. So we still only go into this logic if they had enabled caching for a bulk endpoint. If that's true, do we crash or default back to sending the response without adding to the cache? I do think that sounds ok.